masq/src/terminal_interface.rs

```diff
@@ -0,0 +1,439 @@
+
+        assert!(
+            given_output.contains(&"A".repeat(90)),
+            "synchronized: {}",
+            given_output
+        );
+        assert!(
+            given_output.contains(&"B".repeat(90)),
+            "synchronized: {}",
+            given_output
+        );
+    }
+
+    fn test_terminal_collision<C>(closure1: Box<C>, closure2: Box<C>) -> String
+    where
+        C: FnMut(TerminalWrapper, MixingStdout) -> () + Sync + Send + 'static,
+    {
+        let interface = TerminalWrapper::new(Box::new(TerminalActiveMock::new()));
+
+        let barrier = Arc::new(Barrier::new(2));
+
+        let (tx, rx) = unbounded();
+        let stdout_c1 = MixingStdout::new(tx);
+        let stdout_c2 = stdout_c1.clone();
+
+        let handles: Vec<_> = vec![(closure1, stdout_c1), (closure2, stdout_c2)]
+            .into_iter()
+            .map(|pair| {
+                let (mut closure, stdout): (Box<C>, MixingStdout) = pair;
+                let barrier_handle = Arc::clone(&barrier);
+                let thread_interface = interface.clone();
+
+                thread::spawn(move || {
+                    barrier_handle.wait();
+                    closure(thread_interface, stdout)
+                })
+            })
+            .collect();
+
+        handles
+            .into_iter()
+            .for_each(|handle| handle.join().unwrap());
+
+        let mut buffer = String::new();
+        loop {
+            match rx.try_recv() {
+                Ok(string) => buffer.push_str(&string),
+                Err(_) => break buffer,
+            }
+        }
+    }
+
+    fn write_in_cycles(written_signal: &str, stdout: &mut dyn Write) {
+        (0..30).for_each(|_| {
+            write!(stdout, "{}", written_signal).unwrap();
+            thread::sleep(Duration::from_millis(1))
+        })
+    }
+
+    #[test]
+    fn configure_interface_complains_that_there_is_no_real_terminal() {
+        let subject = interface_configurator(
+            Box::new(Interface::with_term),
+            Box::new(DefaultTerminal::new),
+        );
+
+        let result = match subject {
+            Ok(_) => panic!("should have been an error, got OK"),
+            Err(e) => e,
+        };
+
+        #[cfg(target_os = "windows")]
+        assert!(result.contains("Local terminal:"), "{}", result);
+        #[cfg(not(windows))]
+        assert!(
+            result.contains("Preparing terminal interface: "),
+            "{}",
+            result
+        );
+        //Windows: The handle is invalid. (os error 6)
+        //Linux: "Getting terminal parameters: Inappropriate ioctl for device (os error 25)"
+    }
+
+    #[test]
+    fn configure_interface_allows_us_starting_in_memory_terminal() {
+        let term_mock = MemoryTerminal::new();
+        let term_mock_clone = term_mock.clone();
+        let terminal_type = move || -> std::io::Result<MemoryTerminal> { Ok(term_mock_clone) };
+        let subject =
```

Comment:
Is this really a subject? It looks like a result. Maybe the subject is actually interface_configurator, which doesn't get a variable.